src/System.Net.Sockets/src/System/Net/Sockets/SocketAsyncEventArgs.Windows.cs

```diff
@@ -155,9 +155,16 @@ private unsafe void PrepareIOCPOperation()
                         "). Returned = " + ((IntPtr)overlapped).ToString("x"));
                 }
             }
-
             Debug.Assert(overlapped != null, "NativeOverlapped is null.");
-            _ptrNativeOverlapped = new SafeNativeOverlapped(_currentSocket.SafeHandle, overlapped);
+
+            if (_ptrNativeOverlapped != null && _ptrNativeOverlapped.SocketHandle == _currentSocket.SafeHandle)
```

Comment:
> Is there a case when _ptrNativeOverlapped.SocketHandle == _currentSocket.SafeHandle is not true?

If the same SocketAsyncEventArgs is used with one socket and then with another.  I noticed we didn't have any tests for this, so I've added one as well.